### ISSUE TITLE
In iOS for Google Login, getAuthorizationCode(...) should not returned JWT but serverAuthCode

### ISSUE THREAD

**Issue by AndrewBargan** [Cap-go/capacitor-social-login#35]
Hi there,

I've noticed that after logging in to Google and retrieving server code like this

```
    await SocialLogin.initialize({
      google: {
        webClientId: "XXXXX",
        iOSClientId: "YYYY",
        iOSServerClientId: "ZZZ"
      },
    });

    const loginRes = await SocialLogin.login({
      provider: "google",
      options: {
        scopes: [
          "profile",
          "email",
          "https://www.googleapis.com/auth/calendar",
          "https://www.googleapis.com/auth/calendar.events",
          "https://www.googleapis.com/auth/calendar.settings.readonly",
        ],
        grantOfflineAccess: true,
      },
    });

    const authorizationCodeRes = await SocialLogin.getAuthorizationCode({ provider: 'google' });
```
Access token is in `loginRes.result.accessToken.token` which is fine but it is equal to `authorizationCodeRes.jwt` (these string values start with 'ya29.a0...'). So `SocialLogin.getAuthorizationCode(...)` should not return JWT which is already accessible in the result of `SocialLogin.login(...)`. It should return `Authorization Code` (its string value starts with '4/0AVG7fi...') which will be exchanged on the Back-end for the `Refresh Token`.

Or you could return refreshToken somehow. We need it on the server side to refresh our accessToken for further requests to Google API

**Comment by riderx** [comment]
Hey Andrew thanks for the feedback the plugin now return accessToken on login as expected, thanks

**Comment by AndrewBargan** [comment]
With accessToken was everything fine. But we also need refreshToken which will be used on the server side to update the accessToken. As you may know accessToken expires and needs to be updated periodically. So could you return refreshToken or maybe server authorization code which will be exchanged for refreshToken on the server side?

**Comment by AndrewBargan** [comment]
As an example
- visit Google API playground here https://developers.google.com/oauthplayground/
- select some API and scope
- click Authorize API

You'll get to the "Exchange authorization code for tokens" page where as you can see "Authorization code" field. So we need your plugin to return this authorization code and our back-end server will fetch Access Token and Refresh Token. Or could you please make plugin return both Access Token (now plugin already returns it) AND Refresh Token(doesn't return)?

**Comment by riderx** [comment]
added

**Comment by AndrewBargan** [comment]
Wow! You are so quick! )

Now works perfectly for iOS!
Could you add for Android too?

Thanks!

**Comment by AndrewBargan** [comment]
Just one more question :)
Now result of SocialLogin.login(...) looks like this
<img width="185" alt="Screenshot 2024-10-25 at 09 22 42" src="https://github.com/user-attachments/assets/a440d445-890a-4668-964e-111e8e3451b0">

What if you move all the inner fields of the result.accessToken object directly into result and rename result.accessToken.token to just accessToken? This way, the final response would look like this:
```
{
  provider: "google",
  result: {
    refreshToken: "1//...",
    accessToken: "ya29...",
    userId: "someId",
    idToken: "eyJhbGci...",
    profile: {
      ...
    }
  }
}
```

**Comment by riderx** [comment]
@AndrewBargan what is missing in Android?

**Comment by AndrewBargan** [comment]
@riderx refreshToken is missing

**Comment by AndrewBargan** [comment]
Hey @riderx,
So could you please add it? )
Thanks!

**Comment by AndrewBargan** [comment]
Hey @riderx,

I've noticed that the outdated CodetrixStudio/CapacitorGoogleAuth plugin only returns a refreshToken on iOS (highlighted in green in the image below), while it returns a serverAuthCode on both platforms (highlighted in red). Given this, I thought that if retrieving a refreshToken on Android isn’t possible, you could simply return the serverAuthCode for both platforms. We would then send this code to our backend, where it could be exchanged for both a refreshToken and an accessToken by making a request to the Google API with the serverAuthCode as a payload. This approach should be sufficient, and in this case, there's no need for you to return a refreshToken on iOS.

iOS
<img width="614" alt="Screenshot 2024-10-28 at 19 42 20" src="https://github.com/user-attachments/assets/c22461fd-0e95-4ec8-8a17-b3d50ded14f7">

Android
<img width="462" alt="Screenshot 2024-10-28 at 19 42 05" src="https://github.com/user-attachments/assets/73f3ab74-b5f4-4898-881a-e0708423600f">

**Comment by riderx** [comment]
yes that where we are going @WcaleNieWolny did a PR in this direction

**Comment by WcaleNieWolny** [comment]
@riderx not quite. `serverAuthCode` is offline access. Offline access is currently **NOT** implemented in the PR.

**Comment by AndrewBargan** [comment]
Hey @WcaleNieWolny,

Could you please implement the serverAuthCode feature? :) I believe offline access to Google accounts would be essential for many users of the plugin.

Thanks!

**Comment by WcaleNieWolny** [comment]
@AndrewBargan No problem!

I started a PR here: https://github.com/Cap-go/capacitor-social-login/pull/44.

Right now, I only implemented offline access for the web app, but it should not take long to implement it for all the platforms.

PS: This `offline access` pr will be blocked until @riderx merges https://github.com/Cap-go/capacitor-social-login/pull/37

**Comment by AndrewBargan** [comment]
Hey @WcaleNieWolny,
Are there any updates? )

**Comment by WcaleNieWolny** [comment]
@AndrewBargan I have fully implemented offline access here: https://github.com/Cap-go/capacitor-social-login/pull/56

This PR is basically ready to be merged. Right now I am attempting to write the docs for it.
Please know, that custom scopes are still not implemented. When I finish writing the docs, I will implement custom scopes

**Comment by AndrewBargan** [comment]
Awesome! Looking forward to being able to download this update! Thanks a lot!

**Comment by WcaleNieWolny** [comment]
@AndrewBargan I just opened a PR [here](https://github.com/Cap-go/capacitor-social-login/pull/58) with the scopes impl for iOS, web

**Comment by WcaleNieWolny** [comment]
@AndrewBargan I just released [0.0.68-alpha.6](https://www.npmjs.com/package/@capgo/capacitor-social-login/v/0.0.68-alpha.6). You can do [offline access](https://github.com/Cap-go/capacitor-social-login/blob/development/docs/setup_google.md#using-offline-access) and you can also specify the scopes.

**Comment by AndrewBargan** [comment]
@WcaleNieWolny Now I'm getting the following error
<img width="1245" alt="Screenshot 2024-11-10 at 21 25 08" src="https://github.com/user-attachments/assets/9df88606-e094-45d0-9b64-239e8ad66141">

and now GoogleLoginOfflineResponse has only 2 fields but could you please return profile data as well?

<img width="405" alt="Screenshot 2024-11-10 at 21 26 40" src="https://github.com/user-attachments/assets/1a442218-8b3e-4b4c-b2f0-7ace64b3adfd">

Like it's done for GoogleLoginOnlineResponse

<img width="404" alt="Screenshot 2024-11-10 at 21 28 23" src="https://github.com/user-attachments/assets/b8f4dc70-be75-4e10-91bb-1dfa02a2dc70">

**Comment by WcaleNieWolny** [comment]
As for this web error - this should not happen. I will fix it tomorrow morning.

As for the profile, no I could not return it from the plugin. It's simply not possible to use server auth code and exchange it for the user data without using a backend.

If you follow the documentation, you will see an example of how you can get the user's profile

**Comment by AndrewBargan** [comment]
@WcaleNieWolny  Got it! Then we'll fetch user profile data on our back-end server. Thanks a lot!
BTW the error occured on Android.

**Comment by WcaleNieWolny** [comment]
@AndrewBargan Hey, I am sorry but I forgot about this bug. As for merging, I have started discussing the merging process with Martin today. He had made some comments about the development branch, and right now we are in a back-and-forth trying to reach a point where he can merge my branch into main.

**Comment by AndrewBargan** [comment]
Sounds good! Thanks!
Have a good day!

**Comment by WcaleNieWolny** [comment]
@AndrewBargan I just pushed https://github.com/Cap-go/capacitor-social-login/commit/f53a7cda5b65bf7947f9a74c7448a97d51385947. This commit should resolve this `profile` problem.

**Comment by AndrewBargan** [comment]
@WcaleNieWolny  Great! Thanks a lot! :)

**Comment by SamX23** [comment]
Hi @WcaleNieWolny does "grantOfflineAccess: true" still working using the latest library (currently: 0.0.81) ?

I also have a case to retrieve an Authorization token, but I'm using the latest(0.0.81) and seems that it keep throwing error on Android.

```
await SocialLogin.login({
  provider: "google",
  options: {
    scopes: [
      "https://www.googleapis.com/auth/calendar",
      "https://www.googleapis.com/auth/meetings.space.created",
    ],
    grantOfflineAccess: true,
  },
});
```

<img width="633" alt="image" src="https://github.com/user-attachments/assets/77c7068f-9ae8-44b5-a85a-6cb3949ba2b2" />

<img width="1183" alt="image" src="https://github.com/user-attachments/assets/010c8946-dfee-4cc6-8b6e-e09205da25fb" />

**Comment by SamX23** [comment]
Using 0.0.77.alpha.1

config:
```
await SocialLogin.initialize({
          google: {
            webClientId: GOOGLE_WEB_CLIENT_ID,
            iOSClientId: GOOGLE_IOS_CLIENT_ID,
            mode: "offline",
          },
        });

        const response = await SocialLogin.login({
          provider: "google",
          options: {
            scopes: [
              "https://www.googleapis.com/auth/calendar",
              "https://www.googleapis.com/auth/meetings.space.created",
            ],
            forceRefreshToken: true,
          },
        });
```


on IOS, I got error of :

```
 {"message":"Cannot find serverAuthCode","errorMessage":"Cannot find serverAuthCode"}
⚡️  [error] - {"message":"Cannot find serverAuthCode","errorMessage":"Cannot find serverAuthCode"}
⚡️  [log] - {"error":{"errorMessage":"Cannot find serverAuthCode"}}
```


on Android, I have build java error:
```
package com.ichowa.app;

import com.getcapacitor.BridgeActivity;
import com.getcapacitor.PluginHandle;

import ee.forgr.capacitor.social.login.GoogleProvider;
import ee.forgr.capacitor.social.login.SocialLoginPlugin;
import ee.forgr.capacitor.social.login.ModifiedMainActivityForSocialLoginPlugin;
import android.content.Intent;

public class MainActivity extends BridgeActivity  implements ModifiedMainActivityForSocialLoginPlugin {
    @Override
    public void onStart() {
        super.onStart();
        bridge.getWebView().setVerticalScrollBarEnabled(false);
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        if (requestCode >= GoogleProvider.REQUEST_AUTHORIZE_GOOGLE_MIN && requestCode < GoogleProvider.REQUEST_AUTHORIZE_GOOGLE_MAX) {
            PluginHandle pluginHandle = getBridge().getPlugin("SocialLogin");
            if (pluginHandle == null) {
                Log.i("Google Activity Result", "SocialLogin login handle is null");
                return;
            }
            Plugin plugin = pluginHandle.getInstance();
            if (!(plugin instanceof SocialLoginPlugin)) {
                Log.i("Google Activity Result", "SocialLogin plugin instance is not SocialLoginPlugin");
                return;
            }
            ((SocialLoginPlugin) plugin).handleGoogleLoginIntent(requestCode, data);
        }
    }

    // This function will never be called, leave it empty
    @Override
    public void IHaveModifiedTheMainActivityForTheUseWithSocialLoginPlugin() {}
}


```

<img width="640" alt="image" src="https://github.com/user-attachments/assets/f9f4cdee-1835-4af1-a023-d1ce60b6b91f" />


---

May I know which version is working for my case which is to use offline mode and custom scope ?

Thanks

**Comment by WcaleNieWolny** [comment]
Try "0.0.69-alpha.9" The latest development version is still in very early beta, whereas with the "0.0.69-alpha.9" version, I am 99.9% sure that it will work well.

**Comment by WcaleNieWolny** [comment]
The setup is slightly different between the latest dev and the old one.

Please follow the old_development docs: https://github.com/Cap-go/capacitor-social-login/blob/old_development/docs/setup_google.md

**Comment by SamX23** [comment]
Thank you so much for the response @WcaleNieWolny !

I've tried it, and yes I can retrieve an Authorization token on Android and IOS on that version, thanks again.

But seems the "0.0.69-alpha.9" does not returning iDtoken on Android and IOS which I also need it for my backend, do you have a suggestion ?

---

<img width="345" alt="image" src="https://github.com/user-attachments/assets/74dc7c8f-0722-442d-9be9-671f3b9e116f" />

---

<img width="255" alt="image" src="https://github.com/user-attachments/assets/465f1aaf-d039-4904-bc76-69ae3a5efe31" />

---

<img width="412" alt="image" src="https://github.com/user-attachments/assets/6cfd3b54-a628-4f50-9a72-8a6e8de5cfa8" />

---

<img width="972" alt="image" src="https://github.com/user-attachments/assets/1e450147-cb9e-4670-8dbb-0079f923236e" />

**Comment by SamX23** [comment]
I see, no worries I will wait and see.

Thank you for the work @WcaleNieWolny !

**Comment by riderx** [comment]
All you need has been merged to master and released, please re open is something is missing
